### ISSUE TITLE
OpenCV 4  dependency with cmake method passes wrong flags to compiler driver

### ISSUE THREAD

**Issue by piotrrak** [mesonbuild/meson#9581]
**Describe the bug**
Recent versions of OpenCV do not provide pkg-config file anymore, and meson uses cmake method to find this dependency.
`${builddir}/build.ninja` contains:
```
build libl.so: cpp_LINKER libl.so.p/dummy.cc.o
 LINK_ARGS = -Wl,--as-needed -Wl,--no-undefined -shared -fPIC -Wl,--start-group -Wl,-soname,libl.so -Wl,--end-group opencv_core opencv_imgcodecs opencv_imgproc
```
Which is wrong, since driver don't know what to do with arguments: `opencv_core opencv_imgcodecs opencv_imgproc`

Not sure if I bark at wrong tree but cmake + gmake works just fine.

**To Reproduce**
Attached small reproduction case that contains two minimal projects: `cuda` and `cpp` with directory `cmake`extracted from my  `${libdir}/cmake/opencv4` 
`$ cd repro_meson_cv45/cpp;  CMAKE_MODULE_PATH=$(pwd)/../cmake meson setup builddir-bad; grep -A1  'build libl.so:' builddir-bad/build.ninja`
or,
`$ cd repro_meson_cv45/cuda; CMAKE_MODULE_PATH=$(pwd)/../cmake meson setup builddir-bad; grep -A1  'build libl.so:' builddir-bad/build.ninja`

**Expected behavior**
Correct build command generated:
```
build libl.so: cpp_LINKER libl.so.p/dummy.cc.o
 LINK_ARGS = -Wl,--as-needed -Wl,--no-undefined -shared -fPIC -Wl,--start-group -Wl,-soname,libl.so -Wl,-lopencv_core -Wl,-lopencv_imgcodecs -Wl,-lopencv_imgproc -Wl,--end-group
```
`{cpp|cuda}/builddir` contains correctly generated files with my workaround (probably wrong)

```
From a1e810082967ac6d70c0630da09d320da03e1116 Mon Sep 17 00:00:00 2001
From: Piotr Rak <piotr.rak@gmail.com>
Date: Tue, 16 Nov 2021 21:00:37 +0100
Subject: [PATCH] DUMB fix for opencv 4.5 cmake dependency

Signed-off-by: Piotr Rak <piotr.rak@gmail.com>
---
 mesonbuild/dependencies/cmake.py | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/mesonbuild/dependencies/cmake.py b/mesonbuild/dependencies/cmake.py
index 5386ef920..71e5edb1f 100644
--- a/mesonbuild/dependencies/cmake.py
+++ b/mesonbuild/dependencies/cmake.py
@@ -521,7 +521,7 @@ class CMakeDependency(ExternalDependency):
             # Try to use old style variables if no module is specified
             if len(libs) > 0:
                 self.compile_args = list(map(lambda x: f'-I{x}', incDirs)) + defs
-                self.link_args = libs
+                self.link_args = list(map(lambda x: f'-Wl,-l{x}', libs))
                 mlog.debug(f'using old-style CMake variables for dependency {name}')
                 mlog.debug(f'Include Dirs:         {incDirs}')
                 mlog.debug(f'Compiler Definitions: {defs}')
-- 
2.25.1
```
Patch also included in attached *.zip file.

**system parameters**
* Is this a [cross build](https://mesonbuild.com/Cross-compilation.html) or just a plain native build (for the same computer)?
NO - it is native build
* what operating system (e.g. MacOS Catalina, Windows 10, CentOS 8.0, Ubuntu 18.04, etc.)
Ubuntu 20.04 LTS
* what Python version are you using e.g. 3.8.0
Python 3.8.10
* what `meson --version`
tested up to 0.60.99 (`git describe 0.60.0-127-g2a0439b6f`)
* what `ninja --version` if it's a Ninja build
1.10.0
* what is `cmake --version`?
cmake version 3.13.5
*what is opencv version?
4.5.3 (`ad6e82942b37be8ee2c71c1d9bc7fe79cd16f7ab`)

**Comment by piotrrak** [comment]
[repro_meson_cv45.zip](https://github.com/mesonbuild/meson/files/7549846/repro_meson_cv45.zip)

**Comment by eli-schwartz** [comment]
> Recent versions of OpenCV do not provide pkg-config file anymore

You need to configure opencv with `-DOPENCV_GENERATE_PKGCONFIG=ON` due to irrational fears.

...

Since this code is part of "Try to use old style variables if no module is specified", I assume that opencv is using "old style variables" which should really be a warning. I doubt unconditionally adding `-l` is the correct generic solution to handle other projects gracefully too... does anyone know how cmake actually expects this to be used? :o

**Comment by piotrrak** [comment]
Thanks for hint, will try again tomorrow.
I guessed it's most likely wrong solution, made this patch just to create correct `build.ninja` for myself for now.

**Comment by mensinda** [comment]
Should be fixed with #9665

**Comment by piotrrak** [comment]
Thanks
